File: code-tiny/tools/project_topology/parsers/manifest.py
"""Safe identity-level adapters for common ecosystem manifests."""

from __future__ import annotations

import json
import re
from typing import Any, Dict

from ..models import (
    AnalysisDiagnostic,
    ConfidenceLevel,
    DescriptorFact,
    DescriptorParseOutput,
    DescriptorRole,
    DescriptorType,
    DiagnosticCode,
    ParseDepth,
    safe_summary,
)
from .common import evidence, module_path_for_file


_NAME_PATTERNS = (
    re.compile(r'(?m)^\s*(?:name|module|package)\s*[=:]\s*["\']?([^"\'\s,}]+)'),
    re.compile(r'(?m)^\s*module\s+([^\s]+)'),
)


def parse_identity_manifest(
    *,
    project_id: str,
    path: str,
    text: str,
    parser: str = "manifest",
    role: DescriptorRole = DescriptorRole.IDENTITY,
    parse_depth: ParseDepth = ParseDepth.IDENTITY,
    secret_bearing: bool = False,
    generated: bool = False,
) -> DescriptorParseOutput:
    module_path = module_path_for_file(path)
    diagnostics = []
    properties: Dict[str, Any] = {}
    name = ""
    if path.lower().endswith(".json"):
        try:
            payload = json.loads(text)
            if isinstance(payload, dict):
                name = str(payload.get("name") or payload.get("module") or "")
                for key in ("version", "type", "private", "workspaces"):
                    if key in payload and key != "workspaces":
                        properties[key] = payload[key]
                if isinstance(payload.get("workspaces"), list):
                    properties["declared_modules"] = sorted(
                        str(item) for item in payload["workspaces"] if isinstance(item, str)
                    )
        except (ValueError, TypeError) as exc:
            diagnostics.append(
                AnalysisDiagnostic(
                    DiagnosticCode.MALFORMED_DESCRIPTOR,
                    f"Malformed JSON manifest: {exc}",
                    file_path=path,
                    module_path=module_path,
                )
            )
    if not name:
        for pattern in _NAME_PATTERNS:
            match = pattern.search(text)
            if match:
                name = match.group(1)
                break
    descriptor = DescriptorFact.create(
        project_id=project_id,
        module_path=module_path,
        path=path,
        descriptor_type=DescriptorType.PACKAGE_MANIFEST,
        role=role,
        parser=parser,
        parse_depth=parse_depth,
        generated=generated,
        canonical=not generated,
        secret_bearing=secret_bearing,
        redacted=secret_bearing,
        summary=(
            "[redacted configuration]"
            if secret_bearing
            else safe_summary(f"{parser} manifest {name or path}")
        ),
        properties={"name": name, **properties},
        confidence=ConfidenceLevel.MEDIUM if diagnostics else ConfidenceLevel.HIGH,
        evidence=evidence(path),
        diagnostics=tuple(diagnostics),
    )
    return DescriptorParseOutput(
        descriptor=descriptor,
        diagnostics=tuple(diagnostics),
    )

File: rust/crates/analyzer-topology/src/parsers/manifest.rs
//! Port `parsers/manifest.py` — safe identity-level adapters cho ecosystem
//! manifests (JSON name extraction + name-pattern regex).

use std::sync::LazyLock;

use regex::Regex;

use crate::models::{confidence, descriptor_type, safe_summary, AnalysisDiagnostic, DescriptorFact, PyValue};
use crate::parsers::common::{evidence, module_path_for_file};
use crate::parsers::DescriptorParseOutput;

static NAME_PATTERN_1: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?m)^\s*(?:name|module|package)\s*[=:]\s*["']?([^"'\s,}]+)"#).expect("regex")
});
static NAME_PATTERN_2: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?m)^\s*module\s+([^\s]+)").expect("regex"));

/// `parse_identity_manifest` — parser/role/depth/flags truyền từ registry spec.
#[allow(clippy::too_many_arguments)]
pub fn parse_identity_manifest(
    project_id: &str,
    path: &str,
    text: &str,
    parser: &str,
    role: &'static str,
    parse_depth_value: &'static str,
    secret_bearing: bool,
    generated: bool,
) -> DescriptorParseOutput {
    let module_path = module_path_for_file(path);
    let mut diagnostics: Vec<AnalysisDiagnostic> = Vec::new();
    let mut properties = PyValue::dict();
    let mut name = String::new();
    if path.to_lowercase().ends_with(".json") {
        match serde_json::from_str::<serde_json::Value>(text) {
            Ok(payload) if payload.is_object() => {
                let object = payload.as_object().expect("checked object");
                // Python: str(payload.get("name") or payload.get("module") or "")
                // — falsy value (""/0/False/None) rơi xuống key kế.
                let candidate = [object.get("name"), object.get("module")]
                    .into_iter()
                    .flatten()
                    .find(|value| json_truthy(value));
                name = candidate
                    .map(|value| match value {
                        serde_json::Value::String(text) => text.clone(),
                        serde_json::Value::Bool(value) => {
                            if *value { "True".into() } else { "False".into() }
                        }
                        other => other.to_string(),
                    })
                    .unwrap_or_default();
                for key in ["version", "type", "private", "workspaces"] {
                    if key == "workspaces" {
                        continue;
                    }
                    if let Some(value) = object.get(key) {
                        // PyValue: bool/int/str đủ cho các key này.
                        let py = match value {
                            serde_json::Value::Bool(value) => PyValue::Bool(*value),
                            serde_json::Value::Number(number) => PyValue::Int(number.as_i64().unwrap_or(0)),
                            serde_json::Value::String(text) => PyValue::Str(text.clone()),
                            other => PyValue::Str(other.to_string()),
                        };
                        properties.set(key, py);
                    }
                }
                if let Some(serde_json::Value::Array(items)) = object.get("workspaces") {
                    let mut declared: Vec<String> = items
                        .iter()
                        .filter_map(|item| item.as_str().map(str::to_string))
                        .collect();
                    declared.sort();
                    declared.dedup();
                    properties.set(
                        "declared_modules",
                        PyValue::List(declared.into_iter().map(PyValue::Str).collect()),
                    );
                }
            }
            Ok(_) => {}
            Err(error) => {
                diagnostics.push(
                    AnalysisDiagnostic::new(
                        crate::models::diagnostic_code::MALFORMED_DESCRIPTOR,
                        &format!("Malformed JSON manifest: {error}"),
                    )
                    .file_path(path)
                    .module_path(&module_path),
                );
            }
        }
    }
    if name.is_empty() {
        for pattern in [&NAME_PATTERN_1, &NAME_PATTERN_2] {
            if let Some(value) = pattern
                .captures(text)
                .and_then(|caps| caps.get(1))
            {
                name = value.as_str().to_string();
                break;
            }
        }
    }
    // properties = {"name": name, **properties} — name TRƯỚC các key khác.
    let mut final_properties = PyValue::dict();
    final_properties.set("name", PyValue::Str(name.clone()));
    if let PyValue::Dict(entries) = properties {
        for (key, value) in entries {
            final_properties.set(&key, value);
        }
    }
    let summary = if secret_bearing {
        "[redacted configuration]".to_string()
    } else {
        safe_summary(&format!(
            "{parser} manifest {}",
            if name.is_empty() { path } else { name.as_str() }
        ))
    };
    let descriptor = DescriptorFact {
        generated,
        canonical: !generated,
        secret_bearing,
        redacted: secret_bearing,
        ..DescriptorFact::create(
            project_id,
            &module_path,
            path,
            descriptor_type::PACKAGE_MANIFEST,
            role,
            parser,
            parse_depth_value,
            summary,
            final_properties,
            if diagnostics.is_empty() { confidence::HIGH } else { confidence::MEDIUM },
            evidence(path),
            diagnostics.clone(),
        )
        .expect("identity manifest descriptor")
    };
    DescriptorParseOutput {
        descriptor,
        dependencies: vec![],
        endpoints: vec![],
        diagnostics,
    }
}

/// Python truthiness cho JSON payload values (name or module or "").
fn json_truthy(value: &serde_json::Value) -> bool {
    match value {
        serde_json::Value::Null => false,
        serde_json::Value::Bool(value) => *value,
        serde_json::Value::Number(number) => number.as_f64().map(|f| f != 0.0).unwrap_or(false),
        serde_json::Value::String(text) => !text.is_empty(),
        serde_json::Value::Array(items) => !items.is_empty(),
        serde_json::Value::Object(map) => !map.is_empty(),
    }
}
